### PR TITLE
BUI calculations

### LINKED ISSUE #3775: SFMS: Calculate forecast and outlook BUI rasters
**Describe the task**
We want to calculate forecast and outlook BUI rasters to feed into hourly FFMC.

**Acceptance Criteria**
 - [ ] Use today's temp, rh, acc 24 hour precip, to calculate DMC grid, today's temp, rh to calculate DC grid, use both as inputs to calculate BUI, and again verify the same (similar? within a decimal place) BUI output against
     - static calculations
     - input grids for old SFMS BUI (request from geospatial)

**Additional context**
- Previous day DMC, DC, BUI comes from SFMS
- New forecast grid/values is the trigger for updating indices
- As with all indices, noon forecast values (20:00UTC) are used to push out daily indices

- Update Schedule

Image

### PR DESCRIPTION
**Pull request by brettedw**
- Adds job to calculate DMC, DC, and BUI rasters from RDPS & SFMS inputs
  - Run schedule may be worth discussion. Currently runs every 2 hours. There is a need to overwrite data, as more 'up-to-date' weather predictions are released 
- Adds WPSDataset functionality
   - from_array class method
   - multi-dataset context manager to help with testing
- Adds S3 Client class - encapsulating all s3 functionality
- Closes #3775 

# Test Links:
[Landing Page](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/morecast)
[Percentile Calculator](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/percentile-calculator)
[C-Haines](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-4042-e1e498-dev.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

**Comment by dgboss** [comment]
Closing to free up resources

File: api/app/utils/s3.py [hunk 1]
```diff
@@ -47,6 +48,14 @@ async def object_exists_v2(target_path: str):
         return await object_exists(client, bucket, target_path)
 
 
+async def all_objects_exist(*s3_keys: str):
```
Comment (conbrad) [inline api/app/utils/s3.py:51]: I think this is unused and can be deleted

File: api/app/utils/s3.py [hunk 2]
```diff
@@ -10,6 +10,7 @@
 from app import config
 
 logger = logging.getLogger(__name__)
+BUCKET = config.get("OBJECT_STORE_BUCKET")
```
Comment (conbrad) [inline api/app/utils/s3.py:13]: I think this is unused and can be deleted

File: api/app/utils/geospatial.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
 from enum import Enum
 import logging
+import os
```
Comment (dgboss) [inline api/app/utils/geospatial.py:3]: A few unused imports here

File: api/app/utils/s3_client.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        for key in s3_keys:
+            key_exists = await self.object_exists(key)
+            if not key_exists:
+                return False
+        return True
+
+    async def put_object(self, key: str, body: Any):
+        await self.client.put_object(Bucket=self.bucket, Key=key, Body=body)
+
+    async def persist_raster_data(self, temp_dir: str, key: str, transform, projection, values, no_data_value) -> str:
```
Comment (dgboss) [inline api/app/utils/s3_client.py:58]: nit: I feel like the client should just be passed a tif to save and that it shouldn't concern itself with creating a raster from the input parameters.

**Review by dgboss — APPROVED** [review]
Amazing work!

**Review by conbrad — APPROVED** [review]

**Review by conbrad — APPROVED** [review]
Nice work!

**Pull request merged**